src/driver/mod.rs

```diff
@@ -144,40 +144,73 @@ impl Drop for Driver {
+                lc @ Lifecycle::Completed(_) => {
+                    // don't cancel completed items
+                    *cycle = lc;
+                }
+
+                Lifecycle::CompletionList(indices) => {
+                    let mut list = indices.clone().into_list(&mut self.ops.completions);
+                    if !io_uring::cqueue::more(list.peek_end().unwrap().flags) {
+                        // This op is complete. Replace with a null Completed entry
+                        *cycle = Lifecycle::Completed(op::CqeResult {
+                            result: Ok(0),
+                            flags: 0,
+                        });
+                    }
+                }
+
+                _ => {
+                    // All other states need cancelling.
+                }
             }
         }
 
-        // cancel all ops
-        for id in cancellable_ops {
-            unsafe {
-                while self
-                    .uring
-                    .submission()
-                    .push(&AsyncCancel::new(id as u64).build().user_data(u64::MAX))
-                    .is_err()
-                {
-                    self.submit().expect("Internal error when dropping driver");
+        // Submit cancellation for all ops marked Ignored
+        for (id, cycle) in self.ops.lifecycle.iter_mut() {
+            if let Lifecycle::Ignored(..) = cycle {
+                unsafe {
+                    while self
+                        .uring
+                        .submission()
+                        .push(&AsyncCancel::new(id as u64).build().user_data(u64::MAX))
+                        .is_err()
+                    {
+                        self.uring
+                            .submit_and_wait(1)
+                            .expect("Internal error when dropping driver");
+                    }
                 }
             }
         }
 
-        // TODO: add a way to know if a multishot op is done sending completions
-        // SAFETY: this is currently unsound for multishot ops
-        while !self
-            .ops
-            .lifecycle
-            .iter()
-            .all(|(_, cycle)| matches!(cycle, Lifecycle::Completed(_)))
-        {
-            // If waiting fails, ignore the error. The wait will be attempted
-            // again on the next loop.
-            let _ = self.wait();
-            self.tick();
+        // Wait until all in flight cancellations have stopped
+        loop {
+            if self.ops.lifecycle.is_empty() {
+                break;
+            }
+            // Cycles are either all ignored or complete
+            let remove = if let (id, Lifecycle::Completed(..)) =
+                &self.ops.lifecycle.iter().next().unwrap()
+            {
+                Some(*id)
+            } else {
+                None
+            };
+
+            // Remove completed lifecycles from Slab.
+            // This prevents worst case quadtratic processing
```

Comment:
I like this logic. Just a typo in what should be quadratic.

Reply:
Super  picky, while you are at it, maybe no period at end of first sentence? One with a period and one without looks odd.

Reply:
The slab iter processing actually still runs through the whole Vec. And this new logic looks like it could pick up the same id over and over, if our attempt to cancel it doesn't work quickly. I had a comment in the other PR about considering a way of baling out, maybe with an error, or a panic, if progress at canceling had stalled.

Was the goal to avoid a new vec building up with ids to cancel? That had the benefit of just calling cancel once for each but it suffered from having to build up a list first.

I hadn't thought it through fully, but I did wonder why the cancel wasn't issued right there, but maybe having to call tick would mean the slab was changing under us and the borrow checker might even have thought that a bad idea.